pkg/proc/threads.go

```diff
@@ -50,6 +52,17 @@ func (tbe ThreadBlockedError) Error() string {
 	return ""
 }
 
+// CommonThread contains fields used by this package, common to all
+// implementations of the Thread interface.
+type CommonThread struct {
+	returnValues []*Variable
```

Comment:
I don't think we should store this extra state at all, it can be derived. The thread already knows the breakpoint it is stopped at, and we can ask that breakpoint to collect return values. This just gives us one more thing to clear when we continue the process.

Reply:
As for optimizations, the reads of these vars should already be cached in our memory layer.

Reply:
There some problems in doing this:

1. we clear all internal breakpoints before exiting Continue, changing this is pretty complicated since they would still have to be cleared before the next call to Next/Step/StepOut
2. from the point of view of backwards compatibility: we never propagated internal breakpoints to clients, it's probably not a big deal, but this is safer
3. I'd like to use this same interface for the return values of a function call where there is no breakpoint

Reply:
We could put a fake breakpoint on the thread, with a special negative ID (like unrecovered-panic) and marked as UserBreakpoint and save the return values there.

Reply:
That's a good point, I was mistakenly thinking of other breakpoints loading vars that are user set.

Just thinking out loud, tell me if this sounds insane: what if instead of storing _any_ kind of state that we have to manage on the thread or process, we return some sort of global state struct from any function in `proc` which manipulates process state? This would prevent us from ever having to manage or clear any mutable state on any long-lived structure we keep in memory, and enforces boundaries that anything that calls `proc` functions just get state back and can ignore any lower level details that `proc` should encapsulate and hide. Essentially we move `(*Debugger) State()` down a level into proc. the `Debugger` could still handle anything higher-level, but instead of it asking the process for its state when it needs it, the lower layer could just return the information directly for any call that changes process state.

I haven't thought through the full implications of this, but seemed like an interesting enough idea to propose.

Reply:
It would work for this, I think, because it's just something extra that we're returning. But ee're also going to need some place to store the state of ongoing function calls, it can't be a local variable of proc.Continue because proc.Continue can return while an injected function call is still in progress, so it needs to be somewhere in Process/Thread.

Reply:
That's a good point. What do you think of a compromise to do both, or do you think the latter negates the former? I'm just trying to remove as much state to manage as possible.

Reply:
The semantics of Continue/Next/Step/StepOut is "resume the process, wait for a stop, mutate Process to reflect the new state", they do all their communication by mutating Process,  I'd rather not add a second return value just for this special thing. It's going to make the use of Continue more verbose, including in our tests.

Reply:
I'd rather go with either extra state or the fake breakpoint route.

Reply:
I don't want to add a separate value and spread the state around either. I'll keep thinking on this problem and see if there is a sensible refactor here, until then this is fine.